File: frontend/src/types/fleet.ts
export interface Vehicle {
  id: string;
  registration_number: string;
  asset_id: string | null;
  driver_name: string | null;
  driver_phone: string | null;
  capacity_litres: number | null;
  compartments: number | null;
  current_lat: number | null;
  current_lon: number | null;
  last_gps_time: string | null;
  status: string;
  created_at: string;
}

export interface Trip {
  id: string;
  vehicle_id: string;
  origin_terminal_id: string | null;
  destination_name: string;
  destination_lat: number | null;
  destination_lon: number | null;
  status: string;
  loaded_volume_litres: number | null;
  gantry_metered_litres: number | null;
  departure_time: string | null;
  arrival_time: string | null;
  ticket_number: string | null;
  created_at: string;
}

export interface EPod {
  id: string;
  trip_id: string;
  delivered_volume_litres: number;
  receiver_name: string | null;
  delivery_time: string | null;
  is_verified: boolean;
  notes: string | null;
  created_at: string;
}

export interface GeofenceZone {
  id: string;
  name: string;
  zone_type: string;
  center_lat: number;
  center_lon: number;
  radius_meters: number | null;
  polygon: Record<string, unknown> | null;
  is_active: boolean;
  created_at: string;
}

File: backend/app/schemas/fleet.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.core.constants import TripStatus


class VehicleCreate(BaseModel):
    registration_number: str = Field(max_length=50)
    asset_id: uuid.UUID | None = None
    driver_name: str | None = None
    driver_phone: str | None = None
    capacity_litres: float | None = None
    compartments: int | None = None


class VehicleUpdate(BaseModel):
    driver_name: str | None = None
    driver_phone: str | None = None
    capacity_litres: float | None = None
    compartments: int | None = None
    current_lat: float | None = None
    current_lon: float | None = None
    status: str | None = None


class VehicleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, frozen=True)

    id: uuid.UUID
    registration_number: str
    asset_id: uuid.UUID | None
    driver_name: str | None
    driver_phone: str | None
    capacity_litres: float | None
    compartments: int | None
    current_lat: float | None
    current_lon: float | None
    last_gps_time: datetime | None
    status: str
    created_at: datetime


class TripCreate(BaseModel):
    vehicle_id: uuid.UUID
    origin_terminal_id: uuid.UUID | None = None
    destination_name: str = Field(max_length=255)
    destination_lat: float | None = None
    destination_lon: float | None = None
    loaded_volume_litres: float | None = None
    gantry_metered_litres: float | None = None
    ticket_number: str | None = None


class TripUpdate(BaseModel):
    status: TripStatus | None = None
    loaded_volume_litres: float | None = None
    gantry_metered_litres: float | None = None
    departure_time: datetime | None = None
    arrival_time: datetime | None = None
    ticket_number: str | None = None


class TripResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, frozen=True)

    id: uuid.UUID
    vehicle_id: uuid.UUID
    origin_terminal_id: uuid.UUID | None
    destination_name: str
    destination_lat: float | None
    destination_lon: float | None
    status: TripStatus
    loaded_volume_litres: float | None
    gantry_metered_litres: float | None
    departure_time: datetime | None
    arrival_time: datetime | None
    ticket_number: str | None
    created_at: datetime


class EPodCreate(BaseModel):
    trip_id: uuid.UUID
    delivered_volume_litres: float
    receiver_name: str | None = None
    delivery_time: datetime | None = None
    notes: str | None = None


class EPodResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, frozen=True)

    id: uuid.UUID
    trip_id: uuid.UUID
    delivered_volume_litres: float
    receiver_name: str | None
    delivery_time: datetime | None
    is_verified: bool
    notes: str | None
    created_at: datetime


class GeofenceZoneCreate(BaseModel):
    name: str = Field(max_length=255)
    zone_type: str
    center_lat: float
    center_lon: float
    radius_meters: float | None = None
    polygon: dict | None = None


class GeofenceZoneResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, frozen=True)

    id: uuid.UUID
    name: str
    zone_type: str
    center_lat: float
    center_lon: float
    radius_meters: float | None
    polygon: dict | None
    is_active: bool
    created_at: datetime
